controllers/volumereplicationgroup_controller.go

```diff
@@ -1145,6 +1211,105 @@ func (v *VRGInstance) createVR(vrNamespacedName types.NamespacedName, state volr
 	return nil
 }
 
+// namespacedName applies to both VolumeReplication resource and pvc as of now.
+// This is because, VolumeReplication resource for a pvc that is created by the
+// VolumeReplicationGroup has the same name as pvc. But in future if it changes
+// functions to be changed would be processVRAsPrimary(), processVRAsSecondary()
+// to either receive pvc NamespacedName or pvc itself as an additional argument.
+func (v *VRGInstance) selectVolumeReplicationClass(namespacedName types.NamespacedName) (string, error) {
+	className := ""
+
+	if !v.vrcUpdated {
+		if err := v.updateReplicationClassList(); err != nil {
+			v.log.Error(err, "Failed to get VolumeReplicationClass list")
+
+			return className, fmt.Errorf("failed to get VolumeReplicationClass list")
+		}
+
+		v.vrcUpdated = true
+	}
+
+	if len(v.replClassList.Items) == 0 {
+		v.log.Info("No VolumeReplicationClass available")
+
+		return className, fmt.Errorf("no VolumeReplicationClass available")
+	}
+
+	storageClass, err := v.getStorageClass(namespacedName)
+	if err != nil {
+		v.log.Info(fmt.Sprintf("Failed to get the storageclass of pvc %s",
+			namespacedName))
+
+		return className, fmt.Errorf("failed to get the storageclass of pvc %s (%w)",
+			namespacedName, err)
+	}
+
+	for index := range v.replClassList.Items {
+		replicationClass := &v.replClassList.Items[index]
+		if storageClass.Provisioner != replicationClass.Spec.Provisioner {
+			continue
+		}
+
+		schedulingInterval, found := replicationClass.Spec.Parameters["schedulingInterval"]
+		if !found {
+			// schedule not present in parameters of this replicationClass.
+			continue
+		}
+
+		// ReplicationClass that matches both VRG schedule and pvc provisioner
+		if schedulingInterval == v.instance.Spec.SchedulingInterval {
+			className = replicationClass.Name
+
+			break
+		}
+	}
+
+	if className == "" {
+		v.log.Info(fmt.Sprintf("No VolumeReplicationClass found to match provisioner and schedule %s/%s",
+			storageClass.Provisioner, v.instance.Spec.SchedulingInterval))
+
+		return className, fmt.Errorf("no VolumeReplicationClass found to match provisioner and schedule")
+	}
+
+	return className, nil
+}
+
+// if the fetched SCs are stashed, fetching it again for the next PVC can be avoided
+// saving a call to the API server
+func (v *VRGInstance) getStorageClass(namespacedName types.NamespacedName) (*storagev1.StorageClass, error) {
+	pvc := &corev1.PersistentVolumeClaim{}
+
+	for index := range v.pvcList.Items {
+		pvcItem := &v.pvcList.Items[index]
+
+		pvcNamespacedName := types.NamespacedName{Name: pvcItem.Name, Namespace: pvcItem.Namespace}
+		if pvcNamespacedName == namespacedName {
+			pvc = pvcItem
+
+			break
+		}
+	}
+
+	if pvc == nil {
```

Comment:
This statement should never be true. A possible fix would be to change line 1280 to:
`var pvc *corev1.PersistentVolumeClaim`
or change line 1293 to `if *pvc == (corev1.PersistentVolumeClaim{}) {`

Reply:
ACK